pkg/operator/encryption/controllers/migrators/inprocess_processor.go

```diff
@@ -0,0 +1,185 @@
+
+// ListProcessor represents a type that processes resources in parallel.
+// It retrieves resources from the server in batches and distributes among set of workers.
+type ListProcessor struct {
+	concurrency   int
+	workerFn      WorkerFunc
+	dynamicClient dynamic.Interface
+	ctx           context.Context
+}
+
+// NewListProcessor creates a new instance of ListProcessor
+func NewListProcessor(ctx context.Context, dynamicClient dynamic.Interface, workerFn WorkerFunc) *ListProcessor {
+	return &ListProcessor{
+		concurrency:   defaultConcurrency,
+		workerFn:      workerFn,
+		dynamicClient: dynamicClient,
+		ctx:           ctx,
+	}
+}
+
+// Run starts processing all the instance of the given GVR in batches.
+// Note that this operation block until all resources have been process, we can't get the next page or the context has been cancelled
+func (p *ListProcessor) Run(gvr schema.GroupVersionResource) error {
+	var errs []error
+	listPager := pager.New(pager.SimplePageFunc(func(opts metav1.ListOptions) (runtime.Object, error) {
+		for {
+			allResource, err := p.dynamicClient.Resource(gvr).List(opts)
+			if err != nil {
+				klog.Warningf("List of %v failed: %v", gvr, err)
+				if errors.IsResourceExpired(err) {
+					token, err := inconsistentContinueToken(err)
+					if err != nil {
+						return nil, err
+					}
+					opts.Continue = token
+					klog.V(4).Infof("Relisting %v after handling expired token", gvr)
+					continue
+				} else if retryable := canRetry(err); retryable == nil || *retryable == false {
+					return nil, err // not retryable or we don't know. Return error and controller will restart migration.
+				} else {
+					if seconds, delay := errors.SuggestsClientDelay(err); delay {
+						time.Sleep(time.Duration(seconds) * time.Second)
+					}
+					klog.V(4).Infof("Relisting %v after retryable error: %v", gvr, err)
+					continue
+				}
+			}
+
+			migrationStarted := time.Now()
+			klog.V(4).Infof("Migrating %d objects of %v", len(allResource.Items), gvr)
+			err = p.processList(allResource)
+			if err != nil {
+				errs = append(errs, err)
+			}
+			klog.V(4).Infof("Migration of %d objects of %v finished in %v", len(allResource.Items), gvr, time.Now().Sub(migrationStarted))
+
+			allResource.Items = nil // do not accumulate items, this fakes the visitor pattern
+			return allResource, nil // leave the rest of the list intact to preserve continue token
+		}
+	}))
+
+	migrationStarted := time.Now()
+	listPager.FullListIfExpired = false // prevent memory explosion from full list
+	_, listErr := listPager.List(p.ctx, metav1.ListOptions{})
+	klog.V(4).Infof("Migration for %v finished in %v", gvr, time.Now().Sub(migrationStarted))
+	errs = append(errs, listErr)
+	return utilerrors.NewAggregate(errs)
+}
+
+func (p *ListProcessor) processList(l *unstructured.UnstructuredList) error {
+	workCh := make(chan *unstructured.Unstructured, p.concurrency)
+	onWorkerErrorCtx, onWorkerErrorCancel := context.WithCancel(context.Background())
```

Comment:
can we do `ctx := WithCancel(p.ctx)` and only wait for that one?

Reply:
we could but then I think we will not collect all the potential errors as workers are also listening on that channel.